spec-tests/src/test/java/org/candlepin/spec/ExportSpecTest.java

```diff
@@ -0,0 +1,329 @@
+ * Copyright (c) 2009 - 2022 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.spec;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import org.candlepin.dto.api.v1.ConsumerTypeDTO;
+import org.candlepin.dto.api.v1.ProductDTO;
+import org.candlepin.resource.CdnApi;
+import org.candlepin.resource.ConsumerTypeApi;
+import org.candlepin.resource.OwnerContentApi;
+import org.candlepin.resource.OwnerProductApi;
+import org.candlepin.resource.RolesApi;
+import org.candlepin.resource.RulesApi;
+import org.candlepin.resource.UsersApi;
+import org.candlepin.spec.bootstrap.assertions.OnlyInHosted;
+import org.candlepin.spec.bootstrap.client.ApiClient;
+import org.candlepin.spec.bootstrap.client.ApiClients;
+import org.candlepin.spec.bootstrap.client.SpecTest;
+import org.candlepin.spec.bootstrap.client.StandardExportFactory;
+import org.candlepin.spec.bootstrap.client.api.ConsumerClient;
+import org.candlepin.spec.bootstrap.client.api.JobsClient;
+import org.candlepin.spec.bootstrap.client.api.OwnerClient;
+import org.candlepin.spec.bootstrap.data.util.FileUtil;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import org.apache.commons.codec.binary.Base64;
+import org.apache.commons.collections.CollectionUtils;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+@SpecTest
+class ExportSpecTest {
+    private ConsumerClient consumerClient;
+    private ConsumerTypeApi consumerTypeApi;
+    private OwnerClient ownerClient;
+    private OwnerProductApi ownerProductApi;
+    private UsersApi userApi;
+    private RolesApi rolesApi;
+    private OwnerContentApi ownerContentApi;
+    private CdnApi cdnApi;
+    private RulesApi rulesApi;
+    private JobsClient jobsClient;
+
+    private StandardExportFactory exportFactory;
+    private File export;
+
+    @BeforeAll
+    void beforeAll() throws Exception {
```

Comment:
I've posted this on a few other posts, but global state config is bad for test isolation. In particular this targets things like the factory and consumer UUID at the bottom of this method, but the majority of the clients aren't stateless, which leaks into tests. In this particular case it's *probably* fine since literally every test actually uses the export factory, but the export itself is probably a bit extraneous at this level. Long-term, however, I don't want to get into a case like the ConsumerResource unit tests, which are simultaneously clunky individually due to a bunch of setup, but also interdependent because of their reliance on global test state.

As an aside, the name "client" for the extensions of the API wrappers is very misleading. It makes it look like they encapsulate owner permissions, when it's not. We should probably rename those.

Reply:
Yeah, I was following the pattern of the original ruby tests for how a lot of this was setup. I did see a note on the Exporter class in the candlepin_scenarios mentioning how creating the exports takes a lot of time, and I'm wondering if that is why the pattern of using a single factory and export for all the tests was chosen for export_spec.rb. 

Also, I removed the "client" from the names of the API wrappers throughout!

candlepin_scenarios Exporter comment:
https://github.com/candlepin/candlepin/blob/f4b25d0b9b765280074c600fd30c1c467fff9a8d/spec/candlepin_scenarios.rb#L319-L327

Reply:
Ahh, I misread it as a `@BeforeEach`.

I'm not sure how to feel here, since we still have cases where we create this export, and then it's not used for some of the tests which, instead, create their own export. This feels like a real good place to use the junit 5 nested test class bits, with its own beforeAll to spinup the necessary to create the single export, and then all the individual tests which examine it (and can do so in isolation) can become part of that nested test class. From there, the other tests which either make their own exports can live on in the base/outer test class and operate individually.

This allows us to start isolating actual cases where we have a need for shared resource or state creation, and minimize the frequency at which we toss stuff into the global or class scope.

Reply:
This makes a lot of sense and I like the use of the nested test class here!